core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -403,48 +366,54 @@ public void onFailure(String source, Exception e) {
     }
 
     public static class ShardRoutingEntry extends TransportRequest {
-        ShardRouting shardRouting;
-        ShardRouting sourceShardRouting;
+        ShardId shardId;
+        String allocationId;
+        long primaryTerm;
         String message;
         Exception failure;
 
         public ShardRoutingEntry() {
         }
 
-        ShardRoutingEntry(ShardRouting shardRouting, ShardRouting sourceShardRouting, String message, @Nullable Exception failure) {
-            this.shardRouting = shardRouting;
-            this.sourceShardRouting = sourceShardRouting;
+        ShardRoutingEntry(ShardId shardId, String allocationId, long primaryTerm, String message, @Nullable Exception failure) {
+            this.shardId = shardId;
+            this.allocationId = allocationId;
+            this.primaryTerm = primaryTerm;
             this.message = message;
             this.failure = failure;
         }
 
-        public ShardRouting getShardRouting() {
-            return shardRouting;
+        public boolean matches(ShardRouting shardRouting) {
+            return shardId.equals(shardRouting.shardId()) && shardRouting.assignedToNode() &&
+                allocationId.equals(shardRouting.allocationId().getId());
```

Comment:
can the aid matching the implementation and the rest just assertions ? it should be enough